influxdata/encoder_test.go

```diff
@@ -0,0 +1,442 @@
+}, {
+	testName: "InvalidTagValue",
+	point: Point{
+		Measurement: "m",
+		Tags: []TagKeyValue{{
+			Key:   "x",
+			Value: "",
+		}},
+		Fields: []FieldKeyValue{{
+			Key:   "f",
+			Value: int64(1),
+		}},
+	},
+	expectError: `invalid tag value x=""`,
+}, {
+	testName: "OutOfOrderTag",
+	point: Point{
+		Measurement: "m",
+		Tags: []TagKeyValue{{
+			Key:   "x",
+			Value: "1",
+		}, {
+			Key:   "a",
+			Value: "1",
+		}},
+		Fields: []FieldKeyValue{{
+			Key:   "f",
+			Value: int64(1),
+		}},
+	},
+	expectError: `tag key "a" out of order \(previous key "x"\)`,
+}, {
+	testName: "InvalidFieldKey",
+	point: Point{
+		Measurement: "m",
+		Fields: []FieldKeyValue{{
+			Key:   "",
+			Value: int64(1),
+		}},
+		// Include an explicit timestamp so that we test the path
+		// in Endline that checks lineHasError.
+		Time: "123456",
+	},
+	expectError: `invalid field key ""`,
+}, {
+	testName: "TimeStampTooEarly",
+	point: Point{
+		Measurement: "m",
+		Fields: []FieldKeyValue{{
+			Key:   "x",
+			Value: int64(1),
+		}},
+		Time: "1000-01-01T12:00:00Z",
+	},
+	expectError: `timestamp 1000-01-01T12:00:00Z: line-protocol value out of range`,
+}, {
+	testName: "TimeStampTooLate",
+	point: Point{
+		Measurement: "m",
+		Fields: []FieldKeyValue{{
+			Key:   "x",
+			Value: int64(1),
+		}},
+		Time: "8888-01-01T12:00:00Z",
+	},
+	expectError: `timestamp 8888-01-01T12:00:00Z: line-protocol value out of range`,
+}}
+
+func TestEncoderDataError(t *testing.T) {
+	c := qt.New(t)
+	for _, test := range encoderDataErrorTests {
+		c.Run(test.testName, func(c *qt.C) {
+			var e Encoder
+			e.StartLine("m")
+			e.AddField("f", MustNewValue(int64(1)))
+			c.Assert(e.Err(), qt.IsNil)
+			initialBytes := string(e.Bytes())
+			encodePoint(&e, test.point)
+			c.Assert(e.Err(), qt.ErrorMatches, test.expectError)
+
+			// Check that the original line is still intact without any of
+			// the new one.
+			c.Assert(string(e.Bytes()), qt.Equals, initialBytes)
+
+			// Check that we can add another line OK.
+			e.ClearErr()
+			e.StartLine("n")
+			e.AddField("g", MustNewValue(int64(1)))
+			c.Assert(e.Err(), qt.IsNil)
+			c.Assert(string(e.Bytes()), qt.Equals, "m f=1i\nn g=1i")
+		})
+	}
+}
+
+func BenchmarkEncode(b *testing.B) {
+	ts := time.Now()
+	field1Val := []byte("ds;livjdsflvkfesdvljkdsnbvlkdfsjbldfsjhbdfklsjbvkdsjhbv")
+	field2Val := []byte("12343456")
+	benchmarks := []struct {
+		name   string
+		encode func(b *testing.B, encoder *Encoder)
+	}{{
+		name: "100-points",
+		encode: func(b *testing.B, e *Encoder) {
+			for j := 0; j < 100; j++ {
+				e.StartLine("measurement")
+				e.AddTag("tagfffffffdsffsdfvgdsfvdsfvdsfvd1", "blahblahblah")
+				e.AddTag("uag2", "dfsvdfsvbsdfvs")
+				e.AddTag("zzzzzzzzzz", "fdbgfdbgf")
+				v, err := NewValueFromBytes(String, field1Val)
+				if err != nil {
+					b.Fatal(err)
+				}
+				e.AddField("f", v)
+				v, err = NewValueFromBytes(Int, field2Val)
+				if err != nil {
+					b.Fatal(err)
+				}
```

Comment:
Are you not adding the second value?   Do you want to move the creation of the value outside of the benchmarking code as well, it feels like you're testing more than just the encoding like this.

Reply:
Yup, I want to benchmark the creation of the value too, as that's usually going to be done when encoding line-protocol.

Reply:
:+1: Should there have been an `e.AddField("f2", v)` here though?